docs/basic/quick-sort.md

```diff
@@ -1,43 +1,244 @@
-
-Consider the dividing process of quick sort. After the "division" of quick sort is over, the sequence $A_{p} \cdots A_{r}$ is divided into $A_{p} \cdots A_{q}$ and $A_{q+1} \cdots A_{r}$ . According to the relationship between the number of elements on the left ( $q - p + 1$ ) and the size of k, we can determine whether to solve recursively on the left or only on the right.
+### Implementation (C++) [^ref2]
 
-It can be proved that the expected valye of time complexity is $O(n)$ .
+```cpp
+// C++ Version
+struct Range {
+  int start, end;
+  Range(int s = 0, int e = 0) { start = s, end = e; }
+};
 
-### Reference
+template <typename T>
+void quick_sort(T arr[], const int len) {
+  if (len <= 0) return;
+  Range r[len];
+  int p = 0;
+  r[p++] = Range(0, len - 1);
+  while (p) {
+    Range range = r[--p];
+    if (range.start >= range.end) continue;
+    T mid = arr[range.end];
+    int left = range.start, right = range.end - 1;
+    while (left < right) {
+      while (arr[left] < mid && left < right) left++;
+      while (arr[right] >= mid && left < right) right--;
+      std::swap(arr[left], arr[right]);
+    }
+    if (arr[left] >= arr[range.end])
+      std::swap(arr[left], arr[range.end]);
+    else
+      left++;
+    r[p++] = Range(range.start, left - 1);
+    r[p++] = Range(left + 1, range.end);
+  }
+}
+```
+
+### Implementation (Python) [^ref2]
+
+```python
+# Python Version
+def quick_sort(alist, first, last):
+    if first >= last:
+        return
+    mid_value = alist[first]
+    low = first
+    high = last
+    while low < high:
+        while low < high and alist[high] >= mid_value:
+            high -= 1
+        alist[low] = alist[high]
+        while low < high and alist[low] < mid_value:
+            low += 1
+        alist[high] = alist[low]
+    alist[low] = mid_value
+    quick_sort(alist, first, low - 1)
+    quick_sort(alist, low + 1, last)
+```
+
+## Properties
+
+### Stability
+
+Quicksort is an unstable sorting algorithm.
+
+### Time Complexity
+
+The best-case and average-case time complexity of quicksort is $O(n\log n)$. The worst-case time complexity of it is $O(n^2)$.
+
+## Optimizations
+
+### Straightforward Optimization 
+
+If we only implement quicksort using basic principles described before (or just copy the straightforward template), then it is barely possible pass [Luogu P1177 Quicksort Template](https://www.luogu.com.cn/problem/P1177), because some toxic data may make time complexity of straightforward implementation $O(n^2)$.
```

Comment:
```suggestion
If we only implement quicksort using basic principles described before (or just copy the straightforward template code), then it is barely possible pass [Luogu P1177 Quicksort Template](https://www.luogu.com.cn/problem/P1177) (Chinese), because some toxic data may make time complexity of straightforward implementation $O(n^2)$.
```